src/amo/reducers/users.js

```diff
@@ -0,0 +1,256 @@
+|};
+
+type FetchUserAccountAction = {|
+  type: typeof FETCH_USER_ACCOUNT,
+  payload: FetchUserAccountParams,
+|};
+
+export const fetchUserAccount = ({
+  errorHandlerId,
+  username,
+}: FetchUserAccountParams): FetchUserAccountAction => {
+  if (!errorHandlerId) {
+    throw new Error('errorHandlerId is required');
+  }
+  if (!username) {
+    throw new Error('username is required');
+  }
+
+  return {
+    type: FETCH_USER_ACCOUNT,
+    payload: {
+      errorHandlerId,
+      username,
+    },
+  };
+};
+
+type LoadCurrentUserAccountParams = {|
+  user: Object,
+|};
+
+type LoadCurrentUserAccountAction = {|
+  type: typeof LOAD_CURRENT_USER_ACCOUNT,
+  payload: LoadCurrentUserAccountParams,
+|};
+
+export const loadCurrentUserAccount = ({
+  user,
+}: LoadCurrentUserAccountParams): LoadCurrentUserAccountAction => {
+  if (!user) {
+    throw new Error('user is required');
+  }
+
+  return {
+    type: LOAD_CURRENT_USER_ACCOUNT,
+    payload: { user },
+  };
+};
+
+type LoadUserAccountParams = {|
+  user: Object,
+|};
+
+type LoadUserAccountAction = {|
+  type: typeof LOAD_USER_ACCOUNT,
+  payload: LoadUserAccountParams,
+|};
+
+export const loadUserAccount = ({
+  user,
+}: LoadUserAccountParams): LoadUserAccountAction => {
+  if (!user) {
+    throw new Error('user is required');
+  }
+
+  return {
+    type: LOAD_USER_ACCOUNT,
+    payload: { user },
+  };
+};
+
+export type LogOutUserAction = {|
+  type: string,
+  payload: Object,
+|};
+
+export function logOutUser(): LogOutUserAction {
+  return {
+    type: LOG_OUT_USER,
+    payload: {},
+  };
+}
+
+export const getCurrentUser = (users: UsersStateType) => {
+  if (!users.currentUserID || !users.byID[users.currentUserID]) {
+    return null;
+  }
+
+  return users.byID[users.currentUserID];
+};
+
+export const getDisplayName = (user: UserType) => {
+  // We fallback to the username if no display name has been defined by the
+  // user.
+  return user.display_name && user.display_name.length ?
+    user.display_name : user.username;
+};
+
+export const hasPermission = (
+  state: { users: UsersStateType }, permission: string,
+): boolean => {
+  const currentUser = getCurrentUser(state.users);
+
+  // If the user isn't authenticated, they have no permissions.
+  if (!currentUser) {
+    return false;
+  }
+
+  const permissions = currentUser.permissions;
+  if (!permissions) {
+    return false;
+  }
+
+  // Admins have absolutely all permissions.
+  if (permissions.includes(ALL_SUPER_POWERS)) {
+    return true;
+  }
+
+  return permissions.includes(permission);
+};
+
+
+type Action =
+  | FetchCurrentUserAccountAction
+  | FetchUserAccountAction
+  | LoadCurrentUserAccountAction
+  | LoadUserAccountAction
+  | LogOutUserAction;
+
+const reducer = (
+  state: UsersStateType = initialState,
+  action: Action
+): UsersStateType => {
+  switch (action.type) {
+    case LOAD_CURRENT_USER_ACCOUNT: {
+      const { user } = action.payload;
+      user.displayName = getDisplayName(user);
```

Comment:
This code is duplicated. You should introduce a helper that both reducer actions can use. Something like:

```js
const { user, newState } = addUserToState({ user: action.payload.user, state });
return {
  ...newState,
  currentUserID: user.id,
};
```